### PR TITLE
OTP Concurrency Updates

### PR DESCRIPTION
**Pull request by ybur-yug**
In reference to issue #939 this removes GenEvent from the lesson and expands on the GenServer portion. It is simply a proposal to promote discussion. Happy to take any feedback and work on it further.

Example of the app from the added portion [here](https://github.com/ybur-yug/genserver_periodic_worker_example).

### CODE REVIEW

**Comment by doomspork** [comment]
I'm not really sure about these changes.  We want to remove GenEvent but this extra example doesn't reenforce anything we covered in the lesson, in fact it covers a callback (`handle_info/2`) that we didn't even talk about.  

If we we're going to expand on the lesson we should do so with the existing example.  What I think we should to do:

+ Cover start up and using our GenServer with the existing example
+ Add a section on `handle_info/2` and we expand our Queue example to support data expiration and using the `perform_after/3` to look for and remove expired data.

**Comment by doomspork** [comment]
Accidental close!  Sorry :grimacing:

File: lessons/advanced/otp-concurrency.md [hunk 1]
```diff
@@ -149,79 +147,120 @@ iex> SimpleQueue.queue
...
+$ iex -S mix
+iex(1)>
+20:39:27.732 [info]  Job Done
 
-See the official [GenEvent](http://elixir-lang.org/docs/stable/elixir/GenEvent.html#content) documentation for a complete list of callbacks and GenEvent functionality.
+20:39:28.232 [info]  Job Done
+```
+
+Thats it!
+Check out a working example of the CronJob app [here](https://github.com/ybur-yug/genserver_periodic_worker_example).
```
Comment (doomspork) [inline lessons/advanced/otp-concurrency.md]: We have refrained from linking to any individual's repos and I'd like to continue to do that.

File: lessons/advanced/otp-concurrency.md [hunk 2]
```diff
@@ -149,79 +147,120 @@ iex> SimpleQueue.queue
 [1, 2, 3, 20]
 ```
 
-For more information check out the official [GenServer](http://elixir-lang.org/docs/stable/elixir/GenServer.html#content) documentation.
+### A Simple Cron Server
```
Comment (doomspork) [inline lessons/advanced/otp-concurrency.md]: Let's discuss how we could expand our existing example project rather than adding a new one.

Comment (ybur-yug) [inline lessons/advanced/otp-concurrency.md]: What are you thinking would be good to go beyond it? Thats where I started but I didn't really get any ideas I liked.

Comment (ybur-yug) [inline lessons/advanced/otp-concurrency.md]: Sorry, just saw the above with ideas 👍

File: lessons/advanced/otp-concurrency.md [hunk 3]
```diff
@@ -67,8 +68,6 @@ defmodule SimpleQueue do
 
   def handle_call(:queue, _from, state), do: {:reply, state, state}
 
-  ### Client API / Helper methods
```
Comment (brain-geek) [inline lessons/advanced/otp-concurrency.md]: why not? That's a good practice

File: lessons/advanced/otp-concurrency.md (hunk 2 above)
Comment (brain-geek) [inline lessons/advanced/otp-concurrency.md]: IMO, this sounds rational - current example is for GenEvent, which is being actively deprecated.

File: lessons/advanced/otp-concurrency.md [hunk 4]
```diff
@@ -149,79 +147,120 @@ iex> SimpleQueue.queue
...
 ```
 
-### Calling Handlers
+There are a few pieces here.
+Let's go over them individually.
 
-In addition to `handle_event/2` GenEvents also support `handle_call/2` among other callbacks.  With `handle_call/2` we can handle specific synchronous messages with our handler.
+#### Startup
+We want to take in a given module, function, argument list, and interval to run the task in.
+From here, we start a GenServer with those are the beginning state:
```
Comment (nscyclone) [inline lessons/advanced/otp-concurrency.md]: `with those are the beginning state`→ `using those as the beginning state`

File: lessons/advanced/otp-concurrency.md [hunk 5]
```diff
@@ -149,79 +147,120 @@ iex> SimpleQueue.queue
 [1, 2, 3, 20]
 ```
 
-For more information check out the official [GenServer](http://elixir-lang.org/docs/stable/elixir/GenServer.html#content) documentation.
+### A Simple Cron Server
+Lets say we have a project that needs to regularly run some task then say the work is completed.
+We will give a simple example of how to do this without having to worry about dependencies or any large amounts of complexity.
 
-## GenEvent
+To start, lets make a simple mix project with a supervision tree:
```
Comment (nscyclone) [inline lessons/advanced/otp-concurrency.md]: `lets`→ `let's`

File: lessons/advanced/otp-concurrency.md [hunk 6]
```diff
@@ -149,79 +147,120 @@ iex> SimpleQueue.queue
 [1, 2, 3, 20]
 ```
 
-For more information check out the official [GenServer](http://elixir-lang.org/docs/stable/elixir/GenServer.html#content) documentation.
+### A Simple Cron Server
+Lets say we have a project that needs to regularly run some task then say the work is completed.
```
Comment (nscyclone) [inline lessons/advanced/otp-concurrency.md]: `Lets`→ `Let's`

File: lessons/advanced/otp-concurrency.md [hunk 7]
```diff
@@ -149,79 +147,120 @@ iex> SimpleQueue.queue
...
+```elixir
+  def init({module, function, args, interval}) do
+    schedule_work(interval)
+    {:ok, {module, function, args, interval}}
   end
-end
 ```
 
-### Using GenEvents
+Now, lets create a worker to pass into this worker for once we get it into our tree in `lib/silly_worker.ex`.
```
Comment (nscyclone) [inline lessons/advanced/otp-concurrency.md]: `lets`→ `let's`

File: lessons/advanced/otp-concurrency.md [hunk 8]
```diff
@@ -149,79 +147,120 @@ iex> SimpleQueue.queue
...
-iex> GenEvent.notify(pid, {:msg, "Hello World"})
-Logging new message: Hello World
-Persisting log message: Hello World
+Now if we just run IEx the worker will run that function on our module every second in its own process under the supervision tree:
 
-iex> GenEvent.call(pid, LoggerHandler, :messages)
-["Hello World"]
 ```
+$ iex -S mix
+iex(1)>
```
Comment (nscyclone) [inline lessons/advanced/otp-concurrency.md]: IIRC we remove the line numbers from the `iex` output.

File: lessons/advanced/otp-concurrency.md [hunk 9]
```diff
@@ -149,79 +147,120 @@ iex> SimpleQueue.queue
...
+20:39:27.732 [info]  Job Done
 
-See the official [GenEvent](http://elixir-lang.org/docs/stable/elixir/GenEvent.html#content) documentation for a complete list of callbacks and GenEvent functionality.
+20:39:28.232 [info]  Job Done
+```
+
+Thats it!
```
Comment (nscyclone) [inline lessons/advanced/otp-concurrency.md]: `Thats it!` → `That's it!`

Comment (ybur-yug) [inline lessons/advanced/otp-concurrency.md]: I'm so bad at apostrophes. I need to get a decent spellchecker for Vim that only runs on markdown files.

File: lessons/advanced/otp-concurrency.md (hunk 3 above)
Comment (ybur-yug) [inline lessons/advanced/otp-concurrency.md]: This was an accidental delete, good catch 👍

File: lessons/advanced/otp-concurrency.md (hunk 8 above)
Comment (doomspork) [inline lessons/advanced/otp-concurrency.md]: We do, good catch @nscyclone 👍

**Comment by ybur-yug** [comment]
I'm going to close this as there are more plans for building on the base-lesson rather than diverting and keeping things bite-sized from @doomspork. You can follow the rest the discussion in Issue #939 if interested in contributing/following.

**Pull request closed without merging**